### PR TITLE
#RI-5661 - add rdi auth

### PR DESCRIPTION
**Pull request by AmirAllayarovSofteq**

### CODE REVIEW

File: redisinsight/api/src/modules/rdi/entities/rdi.entity.ts [hunk 1]
```diff
@@ -7,32 +7,20 @@ export class RdiEntity {
...
-  port?: number;
-
   @Expose()
   @Column({ nullable: false })
   name: string;
 
   @Expose()
-  @Column({ nullable: true })
+  @Column({ nullable: false })
   username: string;
```
Comment (ArtemHoruzhenko) [inline redisinsight/api/src/modules/rdi/entities/rdi.entity.ts:20]: Are you sure that username and password are required fields?

File: redisinsight/api/src/modules/rdi/models/rdi.client.metadata.ts [hunk 2]
```diff
@@ -10,6 +10,7 @@ export class RdiClientMetadata {
   sessionMetadata: SessionMetadata;
 
   @IsNotEmpty()
+  @IsOptional()
   @IsString()
-  id: string;
+  id?: string;
```
Comment (ArtemHoruzhenko) [inline redisinsight/api/src/modules/rdi/models/rdi.client.metadata.ts:15]: no! id is mandatory field! Beside that we must add validation for required fields inside clients pool/storage

File: redisinsight/api/src/modules/rdi/models/rdi.ts [hunk 3]
```diff
@@ -85,8 +47,7 @@ export class Rdi {
   @Expose()
   @IsString()
   @IsNotEmpty()
-  @IsOptional()
-  password?: string;
+  password: string;
```
Comment (ArtemHoruzhenko) [inline redisinsight/api/src/modules/rdi/models/rdi.ts:50]: the same. not sure that username and password are required

File: redisinsight/api/src/modules/rdi/providers/rdi.client.storage.ts [hunk 4]
```diff
@@ -1,54 +1,96 @@
...
-        await this.delete(id);
-        return null;
-      }
     }
 
     return client;
   }
 
   async getByMetadata(rdiClientMetadata: RdiClientMetadata): Promise<RdiClient> {
-    return this.clients.get(RdiClient.generateId(rdiClientMetadata));
```
Comment (ArtemHoruzhenko) [inline redisinsight/api/src/modules/rdi/providers/rdi.client.storage.ts:28]: let's add validation of rdiClientMetadata here. We must check that id, sessionMetadata.userId, sessionMetadata,sessionId are defined (may be something else?)

File: redisinsight/api/src/modules/rdi/providers/rdi.client.storage.ts [hunk 5]
```diff
@@ -1,54 +1,96 @@
...
+   * @param client
+   */
   async set(client: RdiClient): Promise<RdiClient> {
-    // todo: client metadata check
-    // todo: existing client check
-    // const existingClient = this.get(client.id);
+    // Additional validation
+    if (
+      !client.id
+      || !client.metadata.sessionMetadata?.sessionId
```
Comment (ArtemHoruzhenko) [inline redisinsight/api/src/modules/rdi/providers/rdi.client.storage.ts:88]: let's add metadata.id validation also

File: redisinsight/api/src/modules/rdi/rdi.service.ts [hunk 6]
```diff
@@ -29,27 +39,71 @@ export class RdiService {
     return rdi;
   }
 
-  async update(id: string, dto: UpdateRdiDto): Promise<Rdi> {
-    return await this.repository.update(id, dto);
+  async update(rdiClientMetadata: RdiClientMetadata, dto: UpdateRdiDto): Promise<Rdi> {
+    // TODO update dto to get only updated fields
+    const model = classToClass(Rdi, dto);
+
```
Comment (ArtemHoruzhenko) [inline redisinsight/api/src/modules/rdi/rdi.service.ts:45]: we can generate id here (if needed) since factory create will require rdi.id in the matadata

**Review by ArtemHoruzhenko — CHANGES_REQUESTED** [review]

File: redisinsight/api/package.json [hunk 7]
```diff
@@ -34,7 +34,8 @@
     "test:api:cov": "nyc --reporter=html --reporter=text --reporter=text-summary yarn run test:api",
     "test:api:ci:cov": "cross-env nyc -r text -r text-summary -r html yarn run test:api --reporter mocha-multi-reporters --reporter-options configFile=test/api/reporters.json && nyc merge .nyc_output ./coverage/test-run-coverage.json",
     "typeorm:migrate": "cross-env NODE_ENV=staging yarn typeorm migration:generate ./migration/migration",
-    "typeorm:run": "yarn typeorm migration:run"
+    "typeorm:run": "yarn typeorm migration:run",
+    "typeorm:stage:run": "cross-env NODE_ENV=staging yarn typeorm migration:run"
```
[the remark this answers is not in the document]
Comment (AmirAllayarovSofteq) [reply redisinsight/api/package.json:38]: Got it.

File: redisinsight/api/src/modules/rdi/entities/rdi.entity.ts (hunk 1 above)
Comment (AmirAllayarovSofteq) [reply redisinsight/api/src/modules/rdi/entities/rdi.entity.ts:20]: Yep. We have the same logic on frontend side. And RDI team confirmed that username and password are mandatory fields. Username can be blank string.

File: redisinsight/api/src/modules/rdi/client/api.rdi.client.ts [hunk 8]
```diff
@@ -1,81 +1,115 @@
...
+    }
 
     if (response.data?.status === RDI_DEPLOY_FAILED_STATUS) {
       throw new RdiPipelineDeployFailedException(undefined, { error: response.data?.error });
     }
   }
 
-  async deployJob(job: RdiJob): Promise<RdiJob> {
+  async deployJob(): Promise<RdiJob> {
     return null;
```
[the remark this answers is not in the document]
Comment (AmirAllayarovSofteq) [reply redisinsight/api/src/modules/rdi/client/api.rdi.client.ts:83]: Yep. Removed.
Thank you

File: redisinsight/api/src/modules/rdi/rdi.service.ts (hunk 6 above)
Comment (AmirAllayarovSofteq) [reply redisinsight/api/src/modules/rdi/rdi.service.ts:45]: Got it, i will add generated id in "create"

File: redisinsight/api/src/modules/rdi/providers/rdi.client.storage.ts (hunk 5 above)
Comment (AmirAllayarovSofteq) [reply redisinsight/api/src/modules/rdi/providers/rdi.client.storage.ts:88]: Lets do it.
Thank you.

File: redisinsight/api/src/modules/rdi/models/rdi.client.metadata.ts [hunk 9]
```diff
@@ -1,7 +1,8 @@
-import { Session, SessionMetadata } from 'src/common/models/session';
+import {Session, SessionMetadata} from 'src/common/models/session';
```
Comment (ArtemHoruzhenko) [inline redisinsight/api/src/modules/rdi/models/rdi.client.metadata.ts:1]: lint error

File: redisinsight/api/src/modules/rdi/rdi.service.ts (hunk 6 above)
Comment (ArtemHoruzhenko) [reply redisinsight/api/src/modules/rdi/rdi.service.ts:46]: correct. not here

**Review by ArtemHoruzhenko — APPROVED** [review]

**Pull request merged**